feat: vm.toString()

odyslam opened the pull request:
<!--
Thank you for your Pull Request. Please provide a description above and review
the requirements below.

Bug fixes and new features should include tests.
-->

## Motivation

Facilitate conversion to strings to be used in other cheat codes (e.g `--ffi`, `--filewrite`).

Some conversions (e.g bytes -> string) are not trivial in native solidity.

## Solution

<!--
Summarize the solution and provide any necessary context needed to understand
the code change.
-->

- Add the cheat code
- Use UIfmt that already exists by moving it to foundry_common

onbjerg commented:
Looks ok to me - wonder why we use `UiFmt` as opposed to the `Display` traits in ethers as a general thought? Also you have some merge conflicts to resolve :smile:

odyslam commented:
For the use of `UIfmt` I would defer to @mattsse

mattsse commented:
> wonder why we use UiFmt as opposed to the Display traits in ethers as a general thought?

mainly because all `Display` impls in eth types truncate... so this way we have control over what's printed

mattsse commented on evm/src/executor/inspector/cheatcodes/util.rs:
```diff
@@ -2,14 +2,15 @@ use super::Cheatcodes;
-use once_cell::sync::Lazy;
+use foundry_common::fmt::*;
 use revm::{CreateInputs, Database, EVMData};
+use std::str;
```
this should not be needed

mattsse commented on evm/src/executor/inspector/cheatcodes/util.rs:
```diff
@@ -71,6 +72,24 @@ pub fn apply<DB: Database>(
             state.labels.insert(inner.0, inner.1.clone());
             Ok(Bytes::new())
         }
+        HEVMCalls::ToString0(inner) => {
+            Ok(ethers::abi::encode(&[Token::String(inner.0.pretty())]).into())
+        }
+        HEVMCalls::ToString2(inner) => {
+            Ok(ethers::abi::encode(&[Token::String(inner.0.pretty())]).into())
+        }
+        HEVMCalls::ToString3(inner) => {
+            Ok(ethers::abi::encode(&[Token::String(inner.0.pretty())]).into())
+        }
+        HEVMCalls::ToString4(inner) => {
+            Ok(ethers::abi::encode(&[Token::String(inner.0.pretty())]).into())
+        }
+        HEVMCalls::ToString5(inner) => {
+            Ok(ethers::abi::encode(&[Token::String(inner.0.pretty())]).into())
+        }
+        HEVMCalls::ToString6(inner) => {
+            Ok(ethers::abi::encode(&[Token::String(inner.0.pretty())]).into())
+        }
```
this is indeed weird, need to debug this in `abigen!`

odyslam replied:
using expand, it indeed produces 0 to 6, skipping 1.

mattsse reviewed (changes requested):
so far so good, looks like imports need to be updated for test to pass

odyslam commented:
@mattsse let's merge this?

mattsse reviewed (changes requested):
lgtm,

needs nightly formatting `cargo +nightly fmt` and clippy fixes

also the missing `HEVMCalls::ToString1` has been fixed on ethers so you can
`cargo update -p ethers` and update the match statement accordingly

odyslam commented:
@mattsse addressed all your comments.  I tried building locally, but I still encountered the issue. 

How can I make sure that I am building with the latest version of ethers? Also, what was the issue? I tried to follow the PR a bit, but I am not certain I fully understand the bug. 

Thanks!

onbjerg commented:
@odyslam Run (in root):

```
cargo update -p ethers
```

After this the bug where `HEVMCalls::ToString1` is skipped is fixed so you need to update your cheatcode stuff

```
cargo +nightly fmt --all
```

And then you'd need clippy fixes.

onbjerg commented:
Pushed the changes so we can get it merged :smile:

mattsse commented on anvil/core/src/eth/transaction/mod.rs:
```diff
@@ -962,7 +962,7 @@ mod tests {
             _ => panic!("Invalid typed transaction"),
         };
-        assert_eq!(tx.input, b"".into());
+        assert_eq!(tx.input, Into::<Bytes>::into(b""));
```
```suggestion
        assert_eq!(tx.input, Bytes::from(b""));
```
